core/src/main/java/org/testcontainers/containers/GenericContainer.java

```diff
@@ -329,42 +331,40 @@ private void applyConfiguration(CreateContainerCmd createCommand) {
     }
 
     /**
-     * Wait until the container has started. The default implementation simply
-     * waits for a port to start listening; subclasses may override if more
-     * sophisticated behaviour is required.
+     * Specify the {@link WaitStrategy} to use to determine if the container is ready.
+     *
+     * @see Wait#defaultWaitStrategy()
+     * @param waitStrategy the WaitStrategy to use
+     * @return the {@code waitStrategy}, for chaining
      */
-    protected void waitUntilContainerStarted() {
-        waitForListeningPort(DockerClientFactory.instance().dockerHostIpAddress(), getLivenessCheckPort());
+    public WaitStrategy waitingFor(WaitStrategy waitStrategy) {
```

Comment:
I even recommend to use self typing:

```
class Base<SELF extends<BASE<SELF>> {
   SELF withSomething() {
      ...
      return getSelf();
   }

   protected SELF getSelf() {
      return (SELF) this;
   }
}
```

```
class SomeChild<SELF extends<SomeChild<SELF>> extends Base<SELF> {
   SELF withProp() {
      ...
      return getSelf();
   }
}
```

And then:
```
new Base()
    .withSomething();
```

```
SomeChild = new SomeChild()
    .withSomething()
    .withProp();
```


We already use self-typing on DockerfileBuilder:
https://github.com/testcontainers/testcontainers-java/blob/master/core/src/main/java/org/testcontainers/images/builder/dockerfile/traits/AddStatementTrait.java